core/include/vecmem/containers/impl/static_array.ipp

```diff
@@ -122,20 +122,141 @@ static_array<T, N>::data(void) const {
     /*
      * Return a pointer to the underlying data, but the elements are const.
      */
-    return v;
+    return m_array;
 }
 
 template <typename T, std::size_t N>
-template <typename std::size_t... Is, typename... Tp>
-VECMEM_HOST_AND_DEVICE static_array<T, N>::static_array(
-    std::index_sequence<Is...>, Tp&&... a) {
-    /*
-     * Construct the array from an enumerated list of elements of arbitrary
-     * size. This is a fold expression which iterates over the elements of
-     * Is (the indices) and Tp (the values), and inserts them into the inner
-     * array using a comma expression.
-     */
-    (static_cast<void>(v[Is] = a), ...);
```

Comment:
What's wrong with this constructor? I don't see why we would need to get rid of it, I think it is even used in traccc.

Reply:
[Fold expressions](https://en.cppreference.com/w/cpp/language/fold) were only introduced with C\+\+17. When I tried to compile the HIP code using the compiler that we use in the CI, I was getting:

```
In file included from /data/ssd-1tb/projects/vecmem/vecmem/tests/hip/test_hip_containers_kernels.hip:19:
In file included from /data/ssd-1tb/projects/vecmem/vecmem/core/include/vecmem/containers/static_array.hpp:305:
/data/ssd-1tb/projects/vecmem/vecmem/core/include/vecmem/containers/impl/static_array.ipp:254:42: warning: pack fold expression is a C++17 extension [-Wc++17-extensions]
    (static_cast<void>(m_array[Is] = a), ...);
                                         ^
```

As I wrote in the description, I had to come up with a C\+\+14 alternative. I was thinking for a while how to do that, but I could only come up with a recursive function call...